### PR TITLE
Want to Develop list

### PR DESCRIPTION
**Pull request by Jamiras**
Expands upon #1815 by adding a Want to Develop Games list, accessible to all full-Developers. It was decided that the feature should not be available to Junior Developers as they should be focused on making the best single set they can, which would lead to becoming a full-Developer and access to the feature.

The "Want to Develop" button is embedded in the Dev panel, next to the "Make Claim" button.
image

Once a claim is made, the developer can view other interested parties via a secondary link in the Dev panel:
image

That link goes to a page similar to the "View Set Requests" page, listing all the developers who have added the game to their "Want to Develop" list. This gives the claimant insight into other devs that they could bounce ideas off of, or inquire about collaborating.

The "View Developer Interest" link is also visible to moderators, even if the game isn't claimed.

While users who are interested in developing for a specific game may be viewed by a moderator, or a user who has claimed the game, the full list associated to an individual developer is only accessible by the developer.

-----

A link to the developer's list has been added to the developer section of the account dropdown.
image

In addition to adding games the developer has interest in making a set for, games with achievements can be added to the list in case the developer wants to revise a set, or work on tickets.

image

Games are never automatically removed from the Want to Develop list. The developer may choose to remove the game from their list once it's claimed, once the set is released, or whenever they decide it's just cluttering up the list.

-----

I had considered adding site notifications/emails for when a Want to Develop game gets claimed, but decided that should wait until we have a more robust collaboration system. When a user claims a game, they may not want other interested developers jumping on the claim, and there currently isn't any way to prevent that. Interested users can still see the claim on the front page, or in the discord #set-claim logs, or even on the Want to Develop list, and engage with the claimant in the same manner they would currently.

### CODE REVIEW

File: app/Helpers/database/set-claim.php [hunk 1]
```diff
@@ -182,7 +182,16 @@ function getClaimData(int $gameID, bool $getFullData = true): array
...
+    if (is_array($gameID)) {
+        if (empty($gameID)) {
+            return [];
+        }
+        $gameIDs = implode(',', $gameID);
+    } else {
+        $gameIDs = $gameID;
+    }
+
+    $query .= " FROM SetClaim sc WHERE sc.GameID IN ($gameIDs)
```
Comment (Jamiras) [inline app/Helpers/database/set-claim.php:194]: Fetching 25 claims in one query took about as long as fetching 2 claims individually.

File: public/gameInfo.php [hunk 2]
```diff
@@ -1031,6 +1047,17 @@ function submitSetRequest(user, gameID) {
 
                 // Display the claims links if not an event game
                 if (!$isEventGame) {
+                    if ($permissions >= Permissions::Developer) {
+                        $gameMetaBindings['developListType'] = UserGameListType::Develop;
+                        echo Blade::render('
+                            <x-game.add-to-list
+                                :gameId="$gameID"
+                                :type="$developListType"
```
Comment (wescopeland) [inline public/gameInfo.php:1055]: It's hard not to notice the new button having a different letter spacing, font size, and height from the other buttons in the devbox. Visually, the difference is striking enough to mistake the new button as being a "primary action" (the new button is nearly 2x the height of the others).

Recommend when the `type` prop is `UserGameListType::Develop`, the button styles change to match the other buttons in the devbox.

Screenshot 2023-09-11 at 6 57 20 PM

File: public/dev-interest.php
Comment (wescopeland) [inline public/dev-interest.php:1]: While as a team we haven't formally stated new pages should be using Laravel MVC, I have a feeling we will later regret adding new raw .php pages like this, even for more trivial features.

#1833 and #1862 both avoid this by:
* Add a new controller to handle the view: https://github.com/RetroAchievements/RAWeb/pull/1862/files#diff-09011a6920853a763a114b2c30c48af5e3352c02b1718f01f6a74b6818bf14f1

* Add a new root blade component for the page which uses `<x-app-layout>` as the root node and the `pageTitle` prop for the `<title>` tag: https://github.com/RetroAchievements/RAWeb/pull/1862/files#diff-fd452dc927a97fa89a704b32391db8d0dd9b0e9a53350a086e196e58b3451450

* Add a new line to _RouteServiceProvider.php_ to hook up the new route (and keep the PHP file extension out of the URL): https://github.com/RetroAchievements/RAWeb/pull/1862/files#diff-4454baa45c0491c9db3f881c7a954ba52eed2713a53009a7d5c972d2f39f2c52R66

Now the back-end logic is separated in a controller away from the UI. The UI doesn't need any `echo` statements and can be 100% Blade:

```php
<x-app-layout pageTitle="{{ $gameTitle['Title'] }} - Developer Interest">
  <h2>List of Interested Users</h2>

  ...
</x-app-layout>
```

I'm happy to help with this conversion in this PR if you would like.

File: public/gameInfo.php (hunk 2 above)
Comment (Jamiras) [reply public/gameInfo.php:1055]: I can fix the font size by removing the `text-sm` from the span block, but everything I try to set for the image sizes cause the image to disappear.

https://github.com/RetroAchievements/RAWeb/blob/4d1acc47288928caadae2c790cdb00dcb23d9b5c/resources/views/platform/components/game/add-to-list.blade.php#L30-L36

image

I don't think that the logic need be conditional. A slightly smaller button in the header would be fine.

File: public/dev-interest.php
Comment (Jamiras) [reply public/dev-interest.php:1]: I'll see what I can do. I'll message you when/if I need help.

File: app/Platform/Controllers/GameDevInterestController.php [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+{
+    public function __invoke(Request $request): View
+    {
+        if (!$request->user()) {
+            abort(401);
+        }
+
+        $permissions = $request->user()->getAttribute('Permissions');
+        if ($permissions < Permissions::Developer) {
+            abort(401);
```
Comment (wescopeland) [inline app/Platform/Controllers/GameDevInterestController.php:24]: nit: This should probably be a 403. The user is authenticated, but they are not authorized.

File: app/Platform/Controllers/GameDevInterestController.php [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        }
+
+        $gameId = (int) $request->route('game');
+        $gameData = getGameData($gameId);
+        if ($gameData === null) {
+            abort(404);
+        }
+
+        if ($permissions < Permissions::Moderator && !hasSetClaimed($request->user()->User, $gameId, true)) {
+            abort(401);
```
Comment (wescopeland) [inline app/Platform/Controllers/GameDevInterestController.php:34]: nit: This should probably be a 403. The user is authenticated, but they are not authorized.

File: public/gameInfo.php (hunk 2 above)
Comment (wescopeland) [reply public/gameInfo.php:1054]: It looks like something in the SVG for the icons is causing the issue. This is where I ended up:

```php
<button 
    id='{{ $type }}-list-button'
    class='btn'
    type='button'
    title='{{ $buttonTooltip }}'
    onClick="toggleListItem('{{ $type }}')"
>
    <div class="flex items-center gap-x-1">
        <div
            id='add-to-{{ $type }}-list'
            class="{{ $addVisibility }} {{ $type === 'play' ? 'w-[18px] h-[18px]' : '' }}"
        >
            @if ($type === 'play')
                <x-icon.add-to-list />
            @else
                <x-fas-plus class="-mt-0.5 w-[12px] h-[12px]" />
            @endif
        </div>

        <div
            id='remove-from-{{ $type }}-list'
            class="{{ $removeVisibility }} {{ $type === 'play' ? 'w-[18px] h-[18px]' : '' }}"
        >
            @if ($type === 'play')
                <x-icon.remove-from-list />
            @else
                <x-fas-check class="-mt-0.5 w-[12px] h-[12px]" />
            @endif
        </div>

        <span>
            {{ __('user-game-list.' . $type . '.name') }}
        </span>
    </div>
</button>
```

Screenshot 2023-09-12 at 7 57 46 PM

**Comment by televandalist** [comment]
`I had considered adding site notifications/emails for when a Want to Develop game gets claimed, but decided that should wait until we have a more robust collaboration system. When a user claims a game, they may not want other interested developers jumping on the claim, and there currently isn't any way to prevent that. Interested users can still see the claim on the front page, or in the discord #set-claim logs, or even on the Want to Develop list, and engage with the claimant in the same manner they would currently.`

Maybe instead of having a single button for making a primary claim, have two? One for an "open" primary claim and another for a "closed" primary claim. If it's open, interested developers would get a notification and would be able to hop aboard for a collaboration claim. If it's closed, then it's basically locked until it's completed, dropped, expired, or set to open.

Just a thought, but I feel that it would work.

**Comment by televandalist** [comment]
Something else that I think would be useful is being able to make notes/comments of some sort on games marked as "want to develop." It would be a good way for someone to express what they'd be interested in doing for the set, whether it's just badges, RP, a couple specific challenges, etc. 

Maybe a comment system for the "View Developer Interest" page?

File: public/gameInfo.php (hunk 2 above)
Comment (Jamiras) [reply public/gameInfo.php:1054]: I think the non-circle icons look find for the Want to Play button and removed the conditional logic. Thanks for your help.

**Review by wescopeland — APPROVED** [review]
Verified locally, everything is working great.

**Pull request merged**